10 Refactor - Modular UI

CThompson01 opened the pull request:
Changed the UIElements library to be more modular and allow for generic UIElement types. This allows the user to create arrays of UI elements and render all of them using a for loop. This also allows users to check for UI input in the same way, and call a callback when input is detected. 

## Changes
* Added generic UIElement type.
* Added generic DrawUIElement function.
* Added generic CheckUIElementInput function.
* Added ButtonPressed function that calls the button callback when input is detected.
* Changed UINumberLabel to hold a pointer to the value it is displaying.
* Changed implementation of UINumberLabel to be a separate struct from UIButton.
* Changed balance to be stored in game state to allow callbacks to alter it.

Closes #10

CThompson01 committed:
Create UIElement Struct

Created generic UIElement struct that can hold other UI elements to allow all elements to be stored in a list.

CThompson01 committed:
Add UI Element Array

Added an array that holds all UI elements to allow for easier rendering.

CThompson01 committed:
Split Button and Label Implementation

Split the implementation of buttons and labels to be two different structs. Number labels now hold a pointer to the value they are labeling to automatically update when the value is changed. Added number label to the DrawUIElement function. Added CreateButtonElement and CreateNumberLabelElement for easier creation of a UIElement for use in an array. Added GetIndexOfKey for easier searching for a specific UIElement in an array. Changed the order of the arguments for CheckButtonInput to make more sense.

CThompson01 committed:
Create Button and Number Label Creators

Added functions to create buttons and number labels to enhance readability. Removed the array of UIElements in the main file, the array functionality is only needed on windows that have large amounts of UI that needs to be looped through. This array will likely be reimplemented on the settings window.

CThompson01 committed:
Start Implementing Button Callbacks

Started implementing button callbacks with test variables and values.

CThompson01 committed:
Fix Balls Unable to Generate When Last Ball is Removed

Fixed a pointer reference to the linked list tail that was causing balls to stop generating when the tail was removed.

CThompson01 committed:
Change Test Names

Changed test names for the game state to descriptive names. Added pointers to the head ball and current ball in the game state. Moved the variable for balance into the game state rather than just being a pointer to balance.

CThompson01 committed:
Implement Generic Input Check

Implemented the generic input check for UIElements. Changed the naming scheme and functionality of the button check functions.

CThompson01 committed:
Change typedef to be Simultaneous with struct Declaration

Changed typedef to be done simultaneously with struct declaration.

bqrichards commented on src/screens/game.c:
```diff
@@ -76,18 +78,43 @@ void draw_zone(int location, int animation_offset, int value) {
+	PBall **balls_tail;
+} GameState;
+
+int dropButtonCallback(GameState *callback_args) {
```
`callback_args` should be renamed `gameState` to match the other instances of the variable.

CThompson01 replied:
Fixed this in the most recent commit.

CThompson01 committed:
Fix dropButtonCallback Argument Name

Fixed the argument name for gameState in dropButtonCallback to match the rest of the instances of the variable.
